### PR TITLE
Fix deadlock in `Wait` function & fix data races in the tests

### LINKED ISSUE #836: Deadlock in sign package tests
# Problem

There is a race condition in `sign` package which can lead to a deadlock.

# Acceptance Criteria

1. Fix tests in `sign` package,
2. Make sure they run fine with `-race`.

# Notes

Captured traceback:
```
goroutine 22 [chan receive, 9 minutes]:
testing.(*T).Run(0xc4204942d0, 0x7b3da9, 0xf, 0xc4205ec580, 0xc420494201)
	/usr/local/go/src/testing/testing.go:790 +0x2fc
github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite.runTests(0xdd6b00, 0xc4204942d0, 0xc420464780, 0xa, 0x10)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite/suite.go:121 +0x9a
github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite.Run(0xc4204942d0, 0xdd15e0, 0xc4204883a0)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite/suite.go:108 +0x5f4
github.com/status-im/status-go/sign.TestPendingRequestsSuite(0xc4204942d0)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests_test.go:30 +0x51
testing.tRunner(0xc4204942d0, 0x9b6790)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de

goroutine 80 [chan send, 9 minutes]:
github.com/status-im/status-go/sign.(*PendingRequests).complete(0xc42028e330, 0xc420210600, 0xe22e60, 0x0, 0x0, 0xdce6a0, 0xc420492350)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests.go:173 +0x184
github.com/status-im/status-go/sign.(*PendingRequests).Wait(0xc42028e330, 0xc4201de180, 0x24, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests.go:117 +0x10f
github.com/status-im/status-go/sign.PendingRequestsSuite.TestWaitTimeout(0xc420292330, 0xc420292340, 0xc4201da3c0, 0xc42028e330)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests_test.go:222 +0x250
github.com/status-im/status-go/sign.(*PendingRequestsSuite).TestWaitTimeout(0xc4204883a0)
	<autogenerated>:1 +0x57
reflect.Value.call(0xc42030a8a0, 0xc4204a6be8, 0x13, 0x87ae81, 0x4, 0xc420045f80, 0x1, 0x1, 0xc4204806b8, 0x87a440, ...)
	/usr/local/go/src/reflect/value.go:434 +0x905
reflect.Value.Call(0xc42030a8a0, 0xc4204a6be8, 0x13, 0xc420480780, 0x1, 0x1, 0x7b3da9, 0xf, 0x0)
	/usr/local/go/src/reflect/value.go:302 +0xa4
github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite.Run.func2(0xc4201da3c0)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/vendor/github.com/stretchr/testify/suite/suite.go:102 +0x243
testing.tRunner(0xc4201da3c0, 0xc4205ec580)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de

goroutine 82 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0xc42028e334, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc42028e330)
	/usr/local/go/src/sync/mutex.go:134 +0xee
sync.(*RWMutex).Lock(0xc42028e330)
	/usr/local/go/src/sync/rwmutex.go:93 +0x2d
github.com/status-im/status-go/sign.(*PendingRequests).tryLock(0xc42028e330, 0xc4201de180, 0x24, 0x0, 0x0, 0x0)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests.go:140 +0x4c
github.com/status-im/status-go/sign.(*PendingRequests).Approve(0xc42028e330, 0xc4201de180, 0x24, 0x87e95b, 0x10, 0x9b6798, 0x0, 0x0, 0x0, 0x0, ...)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests.go:72 +0x185
github.com/status-im/status-go/sign.PendingRequestsSuite.TestWaitTimeout.func1(0xc4200dd300, 0xc420210600)
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests_test.go:218 +0x7e
created by github.com/status-im/status-go/sign.PendingRequestsSuite.TestWaitTimeout
	/home/jenkins/workspace/status-go/status-go-manual/src/github.com/status-im/status-go/sign/pending_requests_test.go:217 +0x21c
```

Quick analysis:

> The problem is probably with `TestWaitTimeout`. What happened is that `Approve` tried to send a data to the channel and `Wait` timed out and also tried to send data to the same channel. As this is 
a buffered channel with size 1 and nothing reads from, a mutex was never released and resulted in a deadlock

### PR DESCRIPTION
**Pull request by mandrigin**
Closes #836

### CODE REVIEW

File: sign/pending_requests.go
Comment (dshulyak) [inline sign/pending_requests.go:119]: what if error is inprogress?
i though that in such case we should cancel the request anyway

Comment (mandrigin) [reply sign/pending_requests.go:119]: `tryLock` can return errors if
a) there is no such request
b) `Approve` is already called and in-progress now
c) another timeout is in progress now

for me, it makes sense, that we only proceed with a timeout if `Approve` wasn't called before the timeout fires

Comment (dshulyak) [reply sign/pending_requests.go:119]: > b) Approve is already called and in-progress now

so for example, if i approve transaction and it makes a network request which takes longer than timeout passed to wait and eventually times out - only caller of the Approve will receive an error?

this might be a problem if caller of Approve is user and caller of the Wait is dapp, or workflow is different?

Comment (mandrigin) [reply sign/pending_requests.go:119]: Nope, because that is an error of a `completeFunc` and it gets propagated to `Wait` too.
But yeah, if the `completeFunc` never ever finishes, `Wait` will lock.

If we don't do that, we will end up in a really weird place, where a request can both succeed (the call gets propagated further) and timeout at the same time.

Comment (dshulyak) [reply sign/pending_requests.go:119]: thanks, now i recall what was the reason for `for { select {}}` in the wait
seems good to assume that once we are in Approve we need to wait until approve will finish

**Review by dshulyak — APPROVED** [review]

**Review by adambabik — APPROVED** [review]
LGTM

Have you checked it runs fine with `-race` flag? It was detecting race conditions before this fix.

**Comment by mandrigin** [comment]
@adambabik yep, I run tests with `-race` just fine now. There are a few races left in e2e/transactions, but I'm fixing them now and will submit a separate PR.

**Pull request merged**
